src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc

```diff
@@ -0,0 +1,443 @@
+        [&, i = idx_t{0}](const Field& sourceField) mutable {
+          execute(sourceField, targetFieldSet[i++]);
+          return;
+        });
+
+      return;
+    }
+
+    //==========================================================================
+    // Class private methods implementation.
+    //==========================================================================
+
+    template<typename fieldType>
+    void StructuredColumnsToStructuredColumns::doExecute(
+      const Field& sourceField, Field& targetField) const {
+
+      // Make Atlas view objects.
+      const auto sourceView = array::make_view<fieldType, 2>(sourceField);
+      auto targetView = array::make_view<fieldType, 2>(targetField);
+
+
+      // Allocate send buffer.
+      const auto nSend = static_cast<size_t>(
+        sendCounts_.back() + sendDisplacements_.back());
+      auto sendBuffer = std::vector<fieldType>(nSend);
+
+      // Allocate receive buffer
+      const auto nRecv = static_cast<size_t>(
+        recvCounts_.back() + recvDisplacements_.back());
+      auto recvBuffer = std::vector<fieldType>(nRecv);
+
+      // The constructor should have made sure that the intersections match up
+      // with counts and displacements vectors.
+
+      //Loop over send intersections and write to send buffer.
+      std::for_each(sendIntersections_.cbegin(), sendIntersections_.cend(),
+        [&, sendBufferIt = sendBuffer.begin()]
+        (const IndexRange& intersection) mutable {
+
+        indexRangeForEach(intersection,
+          [&](const idx_t i, const idx_t j, const idx_t k){
+            const auto iNode = sourceStructuredColumnsPtr_->index(i, j);
+            *sendBufferIt++ = sourceView(iNode, k);
+            return;
+          });
+
+        return;
+
+      });
+
+      // Perform all to all communication.
+      mpi::comm().allToAllv(sendBuffer.data(), sendCounts_.data(),
+        sendDisplacements_.data(), recvBuffer.data(), recvCounts_.data(),
+        recvDisplacements_.data());
+
+
+      //Loop over recv intersections and read from recv buffer.
+      std::for_each(recvIntersections_.cbegin(), recvIntersections_.cend(),
+        [&, recvBufferIt = recvBuffer.cbegin()]
+        (const IndexRange& intersection) mutable {
+
+        indexRangeForEach(intersection,
+          [&](const idx_t i, const idx_t j, const idx_t k){
+            const auto iNode = targetStructuredColumnsPtr_->index(i, j);
+            targetView(iNode, k) = *recvBufferIt++;
+            return;
+          });
+
+        return;
+
+      });
+
+      // By this point, everything should have worked.
+
+      return;
+    }
+
+    //==========================================================================
+    // Helper function definitions.
+    //==========================================================================
+
+    namespace  {
+
+      // Communicate index range of function spaces to all PEs
+      IndexRangeVector getIndexRanges(
```

Comment:
As this function does more, maybe a better name would be getAndCommunicateIndexRanges?